### ISSUE TITLE
Incorrect handling of string variable that starts with '0x'

### ISSUE THREAD

**Issue by rsyring** [pymssql/pymssql#286]
I recently came across this exception from SQLAlchemy in testing:
```
File "pymssql.pyx", line 432, in pymssql.Cursor.execute (pymssql.c:6293)

ProgrammingError: (ProgrammingError) (102, "Incorrect syntax near 'Q1IkOVu'.DB-Lib error message 102, severity 15:\n
General SQL Server error: Check messages from the SQL Server\n"
)

'INSERT INTO [Associates] ([AssociateCode], [AssociateName], [FirstName], [LastName], [Active], tcs_driver_num)
OUTPUT inserted.[AssociateID]
VALUES (%(code)s, %(name)s, %(FirstName)s, %(LastName)s, %(Active)s, %(tcs_driver_num)s)'
{
'code': u'0nKw7K4eh',
'name': '0xQ1IkOVu rCXlj2f3Y',
'FirstName': u'0xQ1IkOVu',
'LastName': u'rCXlj2f3Y',
'tcs_driver_num': None,
'Active': 1}
```

The value for FirstName is  `u'0xQ1IkOVu'`.  But the SQL server says incorrect syntax.  My guess is that somewhere, this string value is being incorrectly intereprtted/inserted as a binary hex value.  We should look into this further.

**Comment by zhang-z** [comment]
I fell into the same trap. In function `_quote_simple_value` of `_mssql.pyx`, there are several lines as:

    if isinstance(value, (str, bytes)):
        if value[0:2] == b'0x':
            return value

If a string starts with "0x", the quote function doesn't quote it. It was added by damoxc in commit 4bf411a, on 21 Jan 2014.

I don't think that's a good idea. Suggest to just remove these lines.

**Comment by dacohenii** [comment]
I, too, fell into this trap while using `_mssql`, when running an update statement with base64-encoded values. I ended up having to build the query string manually (i.e. the non-safe way) and covert it from a binary string.

It seems like it would make more sense if it were changed in one or more of the following ways:
- remove the check entirely, leaving it to the user to explicitly convert it
- convert from string to hex only if using `%X` or `%x` (as defined [here](https://docs.python.org/2/library/string.html#format-specification-mini-language), not when inserting as `%s`
- ensure entire string is valid hex (though this would not be very performant, and could still cause unexpected behavior)
- make it an option you the user needs to set
- some other way I haven't thought of

**Comment by bladams** [comment]
This behavior may allow SQL injection in certain cases. I wrote a test case to demonstrate here: https://github.com/bladams/pymssql/blob/issue-286-hex-strings/tests/test_queries.py#L119

I like the idea of support for formatting types like %d, %x etc. though it looks like that will require a fairly large refactoring of `_substitute_params` and the quote methods. I would be happy to take a stab at that if others would find it useful but I suspect that most would prefer to do the type conversion themselves.

**Comment by rsyring** [comment]
We should see if @damoxc  can tell us what problem he was solving with this addition.  I do see a need, for something like SQLAlchemy, to be able to tell pymssql that the value is binary and leave it up to us to figure out how to get the value in and out.  @bladams could you see how SQLAlchemy is handling binary values w/ pymmsql now and also see how something like psycopg handles binary values for comparison?

**Comment by bladams** [comment]
I did some research and it seems like sqlalchemy and psycopg just use the argument type (i.e. bytes) to indicate how the parameter should be formatted. For all other types pymssql uses the argument type to determine formatting so I don't think relying on the value to determine they type to format as in this solitary case is a good idea.

I would appreciate it if someone could review and give me feedback on or merge my PR as I think this bug is a fairly serious flaw.

**Comment by ramiro** [comment]
@bladams Thanks for your work and for four analysis. Put in that way it doesn't seem correct we are special-casing 0x-prefixed values by peeking at it.

I've merged your PR.

Fixed in f613a55
